File: app/src/main/java/doit/study/droid/fragments/TopicsChooserFragment.java
package doit.study.droid.fragments;

import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.TaskStackBuilder;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

import doit.study.droid.R;
import doit.study.droid.activities.TotalSummaryActivity;
import doit.study.droid.adapters.TopicsAdapter;
import doit.study.droid.app.App;
import doit.study.droid.data.source.Tag;
import doit.study.droid.data.source.local.QuizDatabase;
import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

//import doit.study.droid.activities.InterrogatorActivity;


public class TopicsChooserFragment extends Fragment implements SearchView.OnQueryTextListener{
    private final static boolean DEBUG = false;
    private static final String RECYCLER_LAYOUT_STATE_KEY = "doit.study.droid.fragments.recycler_layout_state_key";
    private TopicsAdapter topicsAdapter;
    private RecyclerView recyclerView;
    private Parcelable savedRecyclerLayoutState;
    private Disposable disposable;

    @Inject
    QuizDatabase quizDatabase;

    public static TopicsChooserFragment newInstance() {
        TopicsChooserFragment fragment = new TopicsChooserFragment();
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        App.getAppComponent().inject(this);
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        if (DEBUG) Timber.d("onCreate");
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_topics_chooser, container, false);
        return v;
    }

    @Override
    public void onViewCreated(final View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        recyclerView = view.findViewById(R.id.topics_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        topicsAdapter = new TopicsAdapter();
        recyclerView.setAdapter(topicsAdapter);

        final FloatingActionButton floatingActionButton = view.findViewById(R.id.commit_button);
        floatingActionButton.setOnClickListener(v -> {
//                Intent intent = new Intent(getContext(), InterrogatorActivity.class);
//                TaskStackBuilder builder = TaskStackBuilder.create(getContext());
//                builder.addNextIntentWithParentStack(intent);
//                builder.startActivities();
        });
        if(savedInstanceState != null) {   // restore scroll position
            if (DEBUG) Timber.d("Restore recycler state");
            savedRecyclerLayoutState = savedInstanceState.getParcelable(RECYCLER_LAYOUT_STATE_KEY);
            recyclerView.getLayoutManager().onRestoreInstanceState(savedRecyclerLayoutState);
        }

        disposable = quizDatabase.getQuizDao().getTagStatistics()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        // onNext
                        tags -> {
                            topicsAdapter.setTags(tags);
                            Timber.d(Arrays.toString(tags.toArray()));
                            },
                        throwable -> {},
                        () -> {}
                );
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater menuInflater) {
        super.onCreateOptionsMenu(menu, menuInflater);
        menuInflater.inflate(R.menu.activity_topic, menu);
        menuInflater.inflate(R.menu.show_total_summary, menu);

        final MenuItem item = menu.findItem(R.id.action_search);
        final SearchView searchView = (SearchView) MenuItemCompat.getActionView(item);
        searchView.setOnQueryTextListener(this);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.select_all:
                setSelectionToAllTags(true);
                return true;
            case R.id.unselect_all:
                setSelectionToAllTags(false);
                return true;
            case R.id.total_summary:
                Timber.d("Start new activity");
                Intent intent = new Intent(getActivity(), TotalSummaryActivity.class);
                TaskStackBuilder builder = TaskStackBuilder.create(getContext());
                builder.addNextIntentWithParentStack(intent);
                builder.startActivities();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void setSelectionToAllTags(boolean checked){
//        for (TagEntity tag: mMasterCopyTags)
//            tag.setChecked(checked);
        topicsAdapter.notifyDataSetChanged();
    }

    @Override
    public void onPause() {
        if (DEBUG) Timber.d("onPause");
        super.onPause();
        List<Tag> tags = topicsAdapter.getTags();
        List<Integer> checkedTagIds = new ArrayList<>();
        for(Tag t: tags){
            if (t.checkedAnyQuestion) {
                checkedTagIds.add(t.getId());
            }
        }
        Completable update = Completable.fromAction(() -> quizDatabase
                .statisticsDao()
                .updateCheckedQuestionsByTags(checkedTagIds.toArray(new Integer[checkedTagIds.size()])));

        Observable.concat(update.toObservable(), quizDatabase.getQuizDao().getTagStatistics().toObservable())
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(
                updated -> Timber.d("updated: " + updated.get(0)),
                error -> Timber.e(error),
                () -> {}

        );
    }


    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        // save scroll position
        if (DEBUG) Timber.d("onSaveInstanceState");
        outState.putParcelable(RECYCLER_LAYOUT_STATE_KEY, recyclerView.getLayoutManager().onSaveInstanceState());
    }




    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        final List<Tag> filteredModel = filter(topicsAdapter.getTags(), newText);
        topicsAdapter.animateTo(filteredModel);
        // don't know why but with scrollToPosition get buggy behavior
        recyclerView.smoothScrollToPosition(0);
        return true;
    }

    private List<Tag> filter (List<Tag> model, String query){
        query = query.toLowerCase();
        final List<Tag> filteredModel = new ArrayList<>();
        for (Tag tag: model){
            final String text = tag.text.toLowerCase();
            if (text.contains(query)){
                filteredModel.add(tag);
            }
        }
        return filteredModel;
    }

}

File: app/src/main/java/doit/study/droid/fragments/TopicsChooserFragment.kt
package doit.study.droid.fragments

import android.content.ContentProviderOperation
import android.content.Intent
import android.database.Cursor
import android.os.Bundle
import android.os.Parcelable
import android.view.*
import androidx.appcompat.widget.SearchView
import androidx.core.app.TaskStackBuilder
import androidx.core.view.MenuItemCompat
import androidx.fragment.app.Fragment
import androidx.loader.app.LoaderManager
import androidx.loader.content.CursorLoader
import androidx.loader.content.Loader
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager.widget.ViewPager
import com.google.android.material.floatingactionbutton.FloatingActionButton
import doit.study.droid.R
import doit.study.droid.activities.InterrogatorActivity
import doit.study.droid.activities.TotalSummaryActivity
import doit.study.droid.adapters.InterrogatorPagerAdapter
import doit.study.droid.adapters.TopicsAdapter
import doit.study.droid.data.QuizProvider
import doit.study.droid.data.Tag
import timber.log.Timber
import java.util.*
import kotlin.collections.LinkedHashMap
import kotlin.collections.MutableMap
import kotlin.collections.component1
import kotlin.collections.component2
import kotlin.collections.filterValues
import kotlin.collections.iterator
import kotlin.collections.set
import kotlin.collections.toList


class TopicsChooserFragment : Fragment(), LoaderManager.LoaderCallbacks<Cursor>, SearchView.OnQueryTextListener {
    private var topicsAdapter: TopicsAdapter? = null
    private var recyclerView: RecyclerView? = null
    private var topics: MutableMap<Int, Tag> = LinkedHashMap()
    // loaders resets state, have to save in var
    private var savedRecyclerLayoutState: Parcelable? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        setupLoaders()
    }

    private fun setupLoaders() {
        loaderManager.initLoader(TAG_LOADER, null, this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_topics_chooser, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setupLayout(view)
        // restore scroll position
        savedInstanceState?.let {
            if (DEBUG) Timber.d("Restore recycler state")
            savedRecyclerLayoutState = it.getParcelable(RECYCLER_LAYOUT_STATE_KEY)
            recyclerView?.layoutManager?.onRestoreInstanceState(savedRecyclerLayoutState)
        }
    }

    private fun setupLayout(view: View){
        recyclerView = view.findViewById(R.id.topics_view)
        recyclerView?.layoutManager = LinearLayoutManager(context)
        topicsAdapter = TopicsAdapter {tag ->
            topics[tag.id] = tag.copy(selected = !tag.selected)
            topicsAdapter?.submitList(topics.values.toList())
        }
        recyclerView?.adapter = topicsAdapter

        val floatingActionButton = view.findViewById(R.id.commit_button) as FloatingActionButton
        floatingActionButton.setOnClickListener {
            navigateToInterrogatorActivity()
        }
    }

    private fun navigateToInterrogatorActivity(){
        val intent = Intent(context, InterrogatorActivity::class.java)
        val builder = TaskStackBuilder.create(context!!)
        builder.addNextIntentWithParentStack(intent)
        builder.startActivities()
    }


    override fun onCreateOptionsMenu(menu: Menu, menuInflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, menuInflater)
        menuInflater.inflate(R.menu.activity_topic, menu)
        menuInflater.inflate(R.menu.show_total_summary, menu)

        val item = menu.findItem(R.id.action_search)
        val searchView = MenuItemCompat.getActionView(item) as SearchView
        searchView.setOnQueryTextListener(this)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.select_all -> {
                setSelectionToAllTags(true)
                return true
            }
            R.id.unselect_all -> {
                setSelectionToAllTags(false)
                return true
            }
            R.id.total_summary -> {
                Timber.d("Start new activity")
                navigateToTotalSummaryActivity()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    private fun navigateToTotalSummaryActivity(){
        val intent = Intent(activity, TotalSummaryActivity::class.java)
        val builder = TaskStackBuilder.create(context!!)
        builder.addNextIntentWithParentStack(intent)
        builder.startActivities()
    }


    private fun setSelectionToAllTags(checked: Boolean) {
        for ((k, v) in topics) {
            topics[k] = v.copy(selected = checked)
        }
        topicsAdapter?.submitList(topics.values.toList())
    }

    override fun onPause() {
        super.onPause()
        topicsAdapter?.let {
            val ops = ArrayList<ContentProviderOperation>()
            createContentProviderOperationBuilder(forSelectedTag = false)?.let {
                ops.add(it.build())
            }
            createContentProviderOperationBuilder(forSelectedTag = true)?.let {
                ops.add(it.build())
            }
            try {
                activity
                        ?.contentResolver
                        ?.applyBatch(QuizProvider.AUTHORITY, ops)
            } catch (e: Exception) {
                Timber.e(e)
            }
        }
    }

    override fun onResume() {
        super.onResume()
        savedRecyclerLayoutState?.let {
            if (DEBUG) Timber.d("Restore layout in loader")
            recyclerView?.layoutManager?.onRestoreInstanceState(savedRecyclerLayoutState)
        }
    }

    private fun createContentProviderOperationBuilder(forSelectedTag: Boolean = false): ContentProviderOperation.Builder? {
        val selection = StringBuilder()

        for (v in topics.values) {
            if (v.selected == forSelectedTag) {
                appendSelection(selection, v.id)
            }
        }

        if (selection.isNotEmpty()) {
            return ContentProviderOperation.newUpdate(QuizProvider.TAG_URI)
                    .withValue(Tag.Table.SELECTED, forSelectedTag)
                    .withSelection(selection.toString(), null)
        }
        return null
    }


    private fun appendSelection(s: StringBuilder, id: Int) {
        if (s.isNotEmpty()) s.append(" OR ")
        s.append(Tag.Table._ID).append(" = ").append(id)
    }


    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        // save scroll position
        if (DEBUG) Timber.d("onSaveInstanceState")
        outState.putParcelable(RECYCLER_LAYOUT_STATE_KEY, recyclerView?.layoutManager?.onSaveInstanceState())
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor> {
        if (DEBUG) Timber.d("onCreateLoader")
        return when (id) {
            TAG_LOADER -> CursorLoader(activity!!, QuizProvider.TAG_URI, null, null, null, null)
            else -> throw Exception("Wrong id for CursorLoader")
        }
    }

    override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor) {
        if (DEBUG) Timber.d("onLoadFinished")
        when (loader.id) {
            TAG_LOADER -> {
                while (data.moveToNext()) {
                    val tag = Tag.newInstance(data)
                    topics[tag.id] = tag
                }
                if (DEBUG) Timber.d("TAG_LOADER Loaded size: %d", topics.size)
                topicsAdapter?.submitList(topics.values.toList())
                savedRecyclerLayoutState?.let {
                    recyclerView?.layoutManager?.onRestoreInstanceState(savedRecyclerLayoutState)
                }
            }
            else -> { }
        }
    }

    override fun onLoaderReset(loader: Loader<Cursor>) {
        if (DEBUG) Timber.d("onLoaderReset")
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        return false
    }

    override fun onQueryTextChange(newText: String): Boolean {
        val filteredTopics = topics.filterValues { it.name.contains(newText, ignoreCase = true) }
        topicsAdapter?.submitList(filteredTopics.values.toList())
        recyclerView?.smoothScrollToPosition(0)
        return true
    }

    companion object {
        private const val DEBUG = false
        private const val RECYCLER_LAYOUT_STATE_KEY = "doit.study.droid.fragments.recycler_layout_state_key"
        private const val TAG_LOADER = 0

        @JvmStatic
        fun newInstance(): TopicsChooserFragment {
            return TopicsChooserFragment()
        }
    }

}
